### PR TITLE
Add ETag validation to appstore requests

### PR DESCRIPTION
**Pull request by rullzer**
* If the ETag if present store it
* If a stored ETag is present then pass it along (with the original
response) to get
* Add tests
* Added files to classmap

### CODE REVIEW

File: lib/private/App/AppStore/Fetcher/Exception/NotModifiedException.php [hunk 1]
```diff
@@ -0,0 +1,7 @@
+<?php
+
+namespace OC\App\AppStore\Fetcher\Exception;
```
Comment (nickvergessen) [inline lib/private/App/AppStore/Fetcher/Exception/NotModifiedException.php]: header missing

**Review by nickvergessen — DISMISSED** [review]

**Review by MorrisJobke — APPROVED** [review]

File: lib/private/App/AppStore/Fetcher/Fetcher.php [hunk 2]
```diff
@@ -62,15 +64,38 @@ public function __construct(IAppData $appData,
...
+	 * @param string $content
+	 *
 	 * @return array
+	 * @throws NotModifiedException
 	 */
-	protected function fetch() {
+	protected function fetch($ETag, $content) {
+		$options = [];
+
+		if ($ETag !== null) {
```
Comment (nickvergessen) [inline lib/private/App/AppStore/Fetcher/Fetcher.php]: docs say string, so `''` ?

File: lib/private/App/AppStore/Fetcher/Fetcher.php [hunk 3]
```diff
@@ -21,6 +21,8 @@
 
 namespace OC\App\AppStore\Fetcher;
 
+use OC\App\AppStore\Fetcher\Exception\NotModifiedException;
```
Comment (nickvergessen) [inline lib/private/App/AppStore/Fetcher/Fetcher.php]: Does not exist anymore?

File: lib/private/App/AppStore/Fetcher/Fetcher.php [hunk 4]
```diff
@@ -62,15 +64,38 @@ public function __construct(IAppData $appData,
 	/**
 	 * Fetches the response from the server
 	 *
+	 * @param string $ETag
+	 * @param string $content
+	 *
 	 * @return array
+	 * @throws NotModifiedException
```
Comment (nickvergessen) [inline lib/private/App/AppStore/Fetcher/Fetcher.php]: Undefined class

File: lib/private/App/AppStore/Fetcher/Fetcher.php [hunk 5]
```diff
@@ -103,7 +136,12 @@ public function get() {
 
 		// Refresh the file content
 		try {
-			$responseJson = $this->fetch();
+			$responseJson = $this->fetch($ETag, $content);
+			$file->putContent(json_encode($responseJson));
+			return json_decode($file->getContent(), true)['data'];
+		} catch (NotModifiedException $e) {
```
Comment (nickvergessen) [inline lib/private/App/AppStore/Fetcher/Fetcher.php]: Undefined class

**Review by nickvergessen — CHANGES_REQUESTED** [review]

**Comment by rullzer** [comment]
@nickvergessen please have another look

**Review by nickvergessen — APPROVED** [review]
works

**Pull request merged**